### PR TITLE
fix(gatsby-transformer-remark): wait for cache promises before returning

### PR DESCRIPTION
**Pull request by pvdz**
The first commit allows us to run markdown in serial mode again (the remark plugin was not awaiting the cache plugins so it was not properly running in serial).

This may also fix a race condition where the same source is parsed twice at the same time, since the cache.set call of the first may not have been completed when the second starts. But that's a very unlike scenario as these caches are more likely to be hit for subsequent builds, not the same session.

The remaining commits refactor the code. I can be okay to drop most of them if people feel it's too much.

### CODE REVIEW

**Comment by pvdz** [comment]
need to fix tests first

**Comment by pvdz** [comment]
Turns out tests were assuming the error case which was the pre-init fake poisoned cache object which was not async and so not a promise and so triggering syntax errors and uncaught promise errors. Fixed.

File: packages/gatsby-transformer-remark/src/extend-node-type.js [hunk 1]
```diff
@@ -144,30 +144,30 @@ module.exports = (
...
+
+      const ASTGenerationPromise = getMarkdownAST(markdownNode)
+      ASTPromiseMap.set(cacheKey, ASTGenerationPromise)
+
+      // Return the promise that will cache the result if good, and deletes the promise from local cache either way
+      // Note that if this cache hits for another parse, it won't have to wait for the cache
+      return ASTGenerationPromise.then(markdownAST =>
+        // Return a promise that waits for the cache to be set, but that does return the generated AST
+        cache.set(cacheKey, markdownAST).then(() => markdownAST)
+      ).finally(() => ASTPromiseMap.delete(cacheKey))
```
Comment (vladar) [inline packages/gatsby-transformer-remark/src/extend-node-type.js:167]: Can we use `await` and `try fnally`? It is a bit inconsistent this way.

**Review by vladar — APPROVED** [review]
Looks good to me 👍 

Just one minor nit inline. Up to you.

File: packages/gatsby-transformer-remark/src/extend-node-type.js (hunk 1 above)
Comment (pvdz) [reply packages/gatsby-transformer-remark/src/extend-node-type.js:167]: yeah good catch

**Review by vladar — APPROVED** [review]

**Pull request merged**